### ISSUE TITLE
Freenom.sh delete others subDomains

### ISSUE THREAD

**Issue by SEMagarinos** [mkorthof/freenom-script#11]
Hi , in freenom.com i have 3 A-Records but when i run 
./freenom.sh -u surink.tk -s www

Delete record “SurInk.tk” and testing.surink.tk

Can you test?

**Comment by mkorthof** [comment]
Hi, this happened on older versions... are you using the latest version?
Does `./freenom.sh -z surink.tk` return any records?
If it still happening you could enable debug and send me the output so i can have a look.

**Comment by SEMagarinos** [comment]
Hi!, I am using the version "`v2019-06-21`" and I show you the captures.

How to switch to debug mode ./freenom.sh?

Yes "-z" return the records. Ex:

`./freenom.sh -z surink.ml`

> DNS Zone: "surink.ml" (1162319329)
> 
> Domain Record: name="SURINK.ML" ttl="300" type="A" value="127.0.0.1"
> Domain Record: name="FRENOMTEST" ttl="3600" type="A" value="127.0.0.1"
> Domain Record: name="PRUEBAS" ttl="3600" type="A" value="127.0.0.1"
> Domain Record: name="WWW" ttl="300" type="A" value="127.0.0.1"

`./freenom.sh -s surink.ml -u WWW`

`./freenom.sh -z surink.ml`

> DNS Zone: "surink.ml" (1162319329)
>
> Domain Record: name="WWW" ttl="3600" type="A" value="181.45.111.17"

**Comment by mkorthof** [comment]
Hmm.. I havent had much time to look at it further but could not replicate. 

The script does check for other records and then calls 'addrecord' instead. 

Debug can be set with debug=1 in conf.

**Comment by SEMagarinos** [comment]
Hi, sorry for the delay

` freenom.sh -z surink.ml`

> 
> Error: Logfile "/var/webs/z-Scripts/freenom/logs/freenom.log" not writable, using "/tmp/freenom.log" instead
> 
> DNS Zone: "surink.ml" (1162319329)
> 
> Domain Record: name="SURINK.ML" ttl="3600" type="A" value="8.8.8.8"
> Domain Record: name="PRUEBAS" ttl="3600" type="A" value="82.8.8.9"
> Domain Record: name="WWW" ttl="3600" type="A" value="181.45.111.17"
> 

Now with de debug=1

> DEBUG: conf scriptConf=/usr/local/bin/freenom.conf
Error: Logfile "/var/webs/z-Scripts/freenom/logs/freenom.log" not writable, using "/tmp/freenom.log" instead
DEBUG: getDomArgs d_args=-z surink.ml arg_domain_name=surink.ml arg_domain_id= arg_subdomain_name=
DEBUG: args 1=-z 2=surink.ml 3= 4= 5= 6=
DEBUG: opts/conf freenom_out_dir=/var/webs/z-Scripts/freenom/logs freenom_out_mask=freenom out_path=/tmp/freenom
DEBUG: opts/conf freenom_domain_name=surink.ml freenom_domain_id= freenom_subdomain_name=
DEBUG: action    freenom_update_ip=0 freenom_update_force=0 freenom_list_records=1
DEBUG: action    freenom_list=0 freenom_list_renewals=0
DEBUG: action    freenom_renew_domain=0 freenom_renew_all=0
DEBUG: myDomains domainId=1162319329 domainName=surink.ml
DEBUG: myDomainsPage match: "surink.ml" = "surink.ml"
> 
> DNS Zone: "surink.ml" (1162319329)
> 
> Domain Record: name="SURINK.ML" ttl="3600" type="A" value="8.8.8.8"
Domain Record: name="PRUEBAS" ttl="3600" type="A" value="82.8.8.9"
Domain Record: name="WWW" ttl="3600" type="A" value="181.45.111.17"


`freenom.sh -u surink.ml -s PRUEBAS`

> DEBUG: conf scriptConf=/usr/local/bin/freenom.conf
> Error: Logfile "/var/webs/z-Scripts/freenom/logs/freenom.log" not writable, using "/tmp/freenom.log" instead
> DEBUG: getDomArgs d_args=-u surink.ml -s PRUEBAS arg_domain_name=surink.ml arg_domain_id= arg_subdomain_name=PRUEBAS
> DEBUG: args 1=-u 2=surink.ml 3=-s 4=PRUEBAS 5= 6=
> DEBUG: opts/conf freenom_out_dir=/var/webs/z-Scripts/freenom/logs freenom_out_mask=freenom out_path=/tmp/freenom
> DEBUG: opts/conf freenom_domain_name=surink.ml freenom_domain_id= freenom_subdomain_name=PRUEBAS
> DEBUG: action    freenom_update_ip=1 freenom_update_force=1 freenom_list_records=0
> DEBUG: action    freenom_list=0 freenom_list_renewals=0
> DEBUG: action    freenom_renew_domain=0 freenom_renew_all=0
> DEBUG: getip i=0 current_ip=181.45.111.17
> DEBUG: myDomains domainId=1162319329 domainName=surink.ml
> DEBUG: myDomainsPage match: "surink.ml" = "surink.ml"
> DEBUG: func_getRec i=0 recType=A recName=SURINK.ML recTTL=3600 recValue=8.8.8.8
> DEBUG: func_getRec i=1 recType=A recName=PRUEBAS recTTL=3600 recValue=82.8.8.9
> DEBUG: func_getRec i=1 type/domain match: recType=A recName=PRUEBAS dnUC=SURINK.ML *OR* sdUC=PRUEBAS
> DEBUG: update_ip i=1 recMatch=1 recordKey=records[1] dnsAction=modify
> DEBUG: update_ip recType=A recName=PRUEBAS recTTL=3600 recValue=82.8.8.9 (empty on add)
> DEBUG: update_ip freenom_update_type=A name=PRUEBAS ttyl=3600 value=181.45.111.17

`freenom.sh -z surink.ml`

> DEBUG: conf scriptConf=/usr/local/bin/freenom.conf
> Error: Logfile "/var/webs/z-Scripts/freenom/logs/freenom.log" not writable, using "/tmp/freenom.log" instead
> DEBUG: getDomArgs d_args=-z surink.ml arg_domain_name=surink.ml arg_domain_id= arg_subdomain_name=
> DEBUG: args 1=-z 2=surink.ml 3= 4= 5= 6=
> DEBUG: opts/conf freenom_out_dir=/var/webs/z-Scripts/freenom/logs freenom_out_mask=freenom out_path=/tmp/freenom
> DEBUG: opts/conf freenom_domain_name=surink.ml freenom_domain_id= freenom_subdomain_name=
> DEBUG: action    freenom_update_ip=0 freenom_update_force=1 freenom_list_records=1
> DEBUG: action    freenom_list=0 freenom_list_renewals=0
> DEBUG: action    freenom_renew_domain=0 freenom_renew_all=0
> DEBUG: myDomains domainId=1162319329 domainName=surink.ml
> DEBUG: myDomainsPage match: "surink.ml" = "surink.ml"
> 
> DNS Zone: "surink.ml" (1162319329)
> 
> Domain Record: name="PRUEBAS" ttl="3600" type="A" value="181.45.111.17"

Any idea?

**Comment by mkorthof** [comment]
Hi,

Judging from the debug output it *should* add change the a record correctly:

<pre>
DEBUG: func_getRec i=0 recType=A recName=SURINK.ML recTTL=3600 recValue=8.8.8.8
DEBUG: func_getRec i=1 recType=A recName=PRUEBAS recTTL=3600 recValue=82.8.8.9
DEBUG: func_getRec i=1 type/domain match: recType=A recName=PRUEBAS dnUC=SURINK.ML OR sdUC=PRUEBAS
DEBUG: update_ip i=1 recMatch=1 recordKey=records[1] dnsAction=<b>modify</b>
DEBUG: update_ip recType=A recName=PRUEBAS recTTL=3600 recValue=82.8.8.9 (empty on add)
DEBUG: update_ip freenom_update_type=A name=PRUEBAS ttyl=3600 value=181.45.111.17
</pre>

It does find the existing record and wants to modify it... 

Each exiting record has a `recordKey` with a number, in your case `records[1]`. You can also see this in the source html code of the dns management page on freenom.com.

If there is no records found the script will do an `addrecord[1]` action, otherwise "modify".
It compares (sub)domain names and if found updates the match.

I've been trying to replicate but in all the tests i did the other records were not touched.

In the newer versions, I've added the record number to debug output - e.g.

```
DEBUG: myDomainsPage match: "example.ml" = "example.ml"
DEBUG: v1=0 v2=type v3=A
DEBUG: v1=0 v2=name v3=EXAMPLE.ML
DEBUG: v1=0 v2=ttl v3=3600
DEBUG: v1=0 v2=value v3=1.2.3.4
DEBUG: v1=1 v2=type v3=A
DEBUG: v1=1 v2=name v3=TEST3
DEBUG: v1=1 v2=ttl v3=3600
DEBUG: v1=1 v2=value v3=1.2.3.4
DEBUG: v1=2 v2=type v3=A
DEBUG: v1=2 v2=name v3=TEST4
DEBUG: v1=2 v2=ttl v3=3600
DEBUG: v1=2 v2=value v3=1.2.3.4
DEBUG: v1=3 v2=type v3=AAAA
DEBUG: v1=3 v2=name v3=WWW
DEBUG: v1=3 v2=ttl v3=3600
DEBUG: v1=3 v2=value v3=2001:1:2:3:4:5:6:7
```

**Comment by SEMagarinos** [comment]
Hi, i try again with other domain.

`freenom.sh -z likem.ml`

> DEBUG: conf scriptConf=/etc/freenom.conf
> 
> DEBUG: getDomArgs d_args=-z likem.ml arg_domain_name=likem.ml arg_domain_id= arg_subdomain_name=
> DEBUG: args 1=-z 2=likem.ml 3= 4= 5= 6= 7= 8= 9=
> DEBUG: opts/conf freenom_out_dir=/var/log/freenom freenom_out_mask=freenom out_path=/var/log/freenom/freenom
> DEBUG: opts/conf freenom_domain_name=likem.ml freenom_domain_id= freenom_subdomain_name=
> DEBUG: action    freenom_update_ip=0 freenom_update_force=0 freenom_list_records=1
> DEBUG: action    freenom_list=0 freenom_list_renewals=0
> DEBUG: action    freenom_renew_domain=0 freenom_renew_all=0
> DEBUG: login cookie_file=/tmp/tmp.42Fp1vLQ4C
> DEBUG: login curl clientarea token=e13c203a0027e4a011b25c9e8cf504faecd9cd1b (r=1/3 http_code=200 errCount=0)
> DEBUG: login curl dologin.php username=usuario@mail.com (r=1/3 http_code=200 errCount=0)
> DEBUG: domains curl myDomainsPage (r=1/3 http_code=200 errCount=0)
> DEBUG: domains curl domainDetails (r=1/3 http_code=200 errCount=0)
> DEBUG: domains domainId=1061180120 domainName=likem.ml
> DEBUG: domains MATCH: "likem.ml" = "likem.ml"
> DEBUG: managedns curl dnsManagementPage (r=1/3 http_code=200 errCount=0)
> 
> DNS Zone: "likem.ml" (1061180120)
> 
> Domain Record: name="GITHUB" ttl="3600" type="A" value="181.45.111.19"
> Domain Record: name="PRUEBAS" ttl="3600" type="A" value="8.8.8.8"
> Domain Record: name="WWW" ttl="3600" type="A" value="181.45.111.17"

Update register A GITHUB 

`freenom.sh -u likem.ml -s GITHUB`

> DEBUG: conf scriptConf=/etc/freenom.conf
DEBUG: getDomArgs d_args=-u likem.ml -s GITHUB arg_domain_name=likem.ml arg_domain_id= arg_subdomain_name=GITHUB
DEBUG: args 1=-u 2=likem.ml 3=-s 4=GITHUB 5= 6= 7= 8= 9=
DEBUG: opts/conf freenom_out_dir=/var/log/freenom freenom_out_mask=freenom out_path=/var/log/freenom/freenom
DEBUG: opts/conf freenom_domain_name=likem.ml freenom_domain_id= freenom_subdomain_name=GITHUB
DEBUG: action    freenom_update_ip=1 freenom_update_force=0 freenom_list_records=0
DEBUG: action    freenom_list=0 freenom_list_renewals=0
DEBUG: action    freenom_renew_domain=0 freenom_renew_all=0
DEBUG: getip i=0 currentIp=181.45.110.56 ipfile=/var/log/freenom/freenom_GITHUB.likem.ml.ip4
DEBUG: login cookie_file=/tmp/tmp.0TtPRT7rOm
DEBUG: login curl clientarea token=44b602a576e756b84dc29dff58df30679c8016ee (r=1/3 http_code=200 errCount=0)
DEBUG: login curl dologin.php username=usuario@mail.com (r=1/3 http_code=200 errCount=0)
DEBUG: domains curl myDomainsPage (r=1/3 http_code=200 errCount=0)
DEBUG: domains curl domainDetails (r=1/3 http_code=200 errCount=0)
DEBUG: domains domainId=1061180120 domainName=likem.ml
DEBUG: domains MATCH: "likem.ml" = "likem.ml"
DEBUG: managedns curl dnsManagementPage (r=1/3 http_code=200 errCount=0)
DEBUG: update_ip func_getRec i=0 recType=A recName=GITHUB recTTL=3600 recValue=181.45.111.19
DEBUG: update_ip func_getRec i=0 type/domain MATCH: recType=A recName=GITHUB -> dnUC=LIKEM.ML *OR* sdUC=GITHUB
DEBUG: update_ip vars - i=0 recMatch=1 recordKey=records[0] dnsAction=modify
DEBUG: update_ip vars - recType=A recName=GITHUB recTTL=3600 recValue=181.45.111.19 (empty on add action)
DEBUG: update_ip vars - freenom_update_type=A name=GITHUB ttyl=3600 value=181.45.110.56
DEBUG: update_ip curl updateResult (r=1/3 http_code=200 errCount=0)
DEBUG: logout curl logoutPage (r=1/3 http_code=200 errCount=0)

and return de error/bug.

`freenom.sh -z likem.ml`

> DEBUG: conf scriptConf=/etc/freenom.conf
> DEBUG: getDomArgs d_args=-z likem.ml arg_domain_name=likem.ml arg_domain_id= arg_subdomain_name=
> DEBUG: args 1=-z 2=likem.ml 3= 4= 5= 6= 7= 8= 9=
> DEBUG: opts/conf freenom_out_dir=/var/log/freenom freenom_out_mask=freenom out_path=/var/log/freenom/freenom
> DEBUG: opts/conf freenom_domain_name=likem.ml freenom_domain_id= freenom_subdomain_name=
> DEBUG: action    freenom_update_ip=0 freenom_update_force=0 freenom_list_records=1
> DEBUG: action    freenom_list=0 freenom_list_renewals=0
> DEBUG: action    freenom_renew_domain=0 freenom_renew_all=0
> DEBUG: login cookie_file=/tmp/tmp.Q4XBoCZRjm
> DEBUG: login curl clientarea token=57fedc661639875bbe0ba1efb03e153df8c8ffce (r=1/3 http_code=200 errCount=0)
> DEBUG: login curl dologin.php username=usuario@mail.com (r=1/3 http_code=200 errCount=0)
> DEBUG: domains curl myDomainsPage (r=1/3 http_code=200 errCount=0)
> DEBUG: domains curl domainDetails (r=1/3 http_code=200 errCount=0)
> DEBUG: domains domainId=1061180120 domainName=likem.ml
> DEBUG: domains MATCH: "likem.ml" = "likem.ml"
> DEBUG: managedns curl dnsManagementPage (r=1/3 http_code=200 errCount=0)
> 
> DNS Zone: "likem.ml" (1061180120)
> 
> Domain Record: name="GITHUB" ttl="3600" type="A" value="181.45.110.56"

**Comment by SEMagarinos** [comment]
I found how you can replicate it.

If DNS records are created from your script, this error does not happen.

But if the record was created from the Freenom page and then run the script, delete the record.

**Comment by mkorthof** [comment]
> I found how you can replicate it.
> If DNS records are created from your script, this error does not happen.
> But if the record was created from the Freenom page and then run the script, delete the record.

Hi, that's awesome you found the cause. Thanks for extensive testing.

I do check the html in dnsmgmt page if there's:
1) any records at all: `addrecord[0]`
2) existing record matches specfied (sub)domain
3) (**change**) record is empty

On _add_ "addrecord[+1]" gets used, else we _modify_ the record

I'll still need test latest changes on domain with records made by Freenom.

Thanks again

**Comment by mkorthof** [comment]
Found what causes this: apparently "addrecord[**+1**]" is not needed and freenom always wants just addrecord[0] and will "shift" existing records itself. It's fixed in next version.

**Comment by mkorthof** [comment]
Fixed in v2019-11-17 (4a8bf3d)
